### PR TITLE
elaborate on the commit messages

### PR DESCRIPTION
**Pull request by greg0ire**
The quality of the history plays a big part on the maintainibility of
any project, so it is important to explain how to do it properly.

### CODE REVIEW

File: doc/CONTRIBUTING.md [hunk 1]
```diff
@@ -72,26 +72,46 @@ If you are not sure of what to do, don't hesitate to open an issue about your PR
...
 
-The commit message has to be clear and related to the PR content.
+Sonata is a big project with many contributors, and a big part of the job is
+being able to understand the code at all times, be it when submitting a PR or
+looking at the history. Good commit messages are crucial to achieve this goal.
 
-The first line of the commit message must be short.
-The other lines must contains a complete description of what you done and why.
+There are already a few articles (or even single purpose websites) about this,
+we cannot recommend enough the following :
```
Comment (soullivaneuh) [inline doc/CONTRIBUTING.md]: `the following :` => `the following:`

File: doc/CONTRIBUTING.md [hunk 2]
```diff
@@ -72,26 +72,46 @@ If you are not sure of what to do, don't hesitate to open an issue about your PR
...
 
-The description is optional but recommended. It could be asked by the team if needed.
+- http://rakeroutes.com/blog/deliberate-git
```
Comment (soullivaneuh) [inline doc/CONTRIBUTING.md]: Dot not know if `-` is more recommended that `*` but I started with `*`, lets keep this way.

Except if there is a good reason of using `-`.

File: doc/CONTRIBUTING.md [hunk 3]
```diff
@@ -72,26 +72,46 @@ If you are not sure of what to do, don't hesitate to open an issue about your PR
...
+- http://stopwritingramblingcommitmessages.com
+- http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
 
-Bad commit message:
+To sum them up, the commit message has to be crystal clear and of course,
+related to the PR content.
+
+The first line of the commit message must be short, keep it under 50
+characters. It must say concisely but *precisely* what you did. The other lines
+must contain a complete description of *why* you did this.
```
Comment (soullivaneuh) [inline doc/CONTRIBUTING.md]: The other lines, if needed, should contains...

In some case for very little modification, having a description is completely trivial.

Example: `Fixing typos on "Get started" documentation` no need more description IMHO.

File: doc/CONTRIBUTING.md (hunk 2 above)
Comment (greg0ire) [inline doc/CONTRIBUTING.md]: Nope, just didn't notice you were using `*`

**Comment by greg0ire** [comment]
@Soullivaneuh fixed

File: doc/CONTRIBUTING.md [hunk 4]
```diff
@@ -72,26 +72,46 @@ If you are not sure of what to do, don't hesitate to open an issue about your PR
...
-Good commit message with description
+The description is optional but strongly recommended. It could be asked by the
+team if needed. PR will often lead to complicated, hard-to-read conversations
+with many links to other web pages. The commit description should be able to
+live without what is said in the PR, and should ideally sum it up in a crystal
+clear way, so that people do not have to open a web browser to understand what
+you did. Links to PR and external references are of course welcome, but should
+not be considered enough.
+
+
```
Comment (soullivaneuh) [inline doc/CONTRIBUTING.md:128]: not needed blank line.

File: doc/CONTRIBUTING.md [hunk 5]
```diff
@@ -72,26 +72,46 @@ If you are not sure of what to do, don't hesitate to open an issue about your PR
...
 
 ```
-Improve search indexing speed for sub-categories
+Document how to install the project
 ```
 
-Good commit message with description
+The description is optional but strongly recommended. It could be asked by the
+team if needed. PR will often lead to complicated, hard-to-read conversations
+with many links to other web pages. The commit description should be able to
```
Comment (soullivaneuh) [inline doc/CONTRIBUTING.md]: The paragraph is too big. Lets place a new line on "The commit description..."

File: doc/CONTRIBUTING.md [hunk 6]
```diff
@@ -72,26 +72,46 @@ If you are not sure of what to do, don't hesitate to open an issue about your PR
...
+live without what is said in the PR, and should ideally sum it up in a crystal
+clear way, so that people do not have to open a web browser to understand what
+you did. Links to PR and external references are of course welcome, but should
```
Comment (soullivaneuh) [inline doc/CONTRIBUTING.md]: - New line on "Links to PR..."
- `to PR` => `to PRs/Issues`

**Comment by greg0ire** [comment]
@Soullivaneuh fixed again

File: doc/CONTRIBUTING.md [hunk 7]
```diff
@@ -72,26 +72,46 @@ If you are not sure of what to do, don't hesitate to open an issue about your PR
...
 ```
 
-Good commit message with description
+The description is optional but strongly recommended. It could be asked by the
+team if needed. PR will often lead to complicated, hard-to-read conversations
+with many links to other web pages.
+
+he commit description should be able to live without what is said in the PR,
+and should ideally sum it up in a crystal clear way, so that people do not have
+to open a web browser to understand what you did. Links to PR and external
```
Comment (soullivaneuh) [inline doc/CONTRIBUTING.md]: You're missing this one:
- New line on "Links to PR..."
- `to PR` => `to PRs/Issues`

File: doc/CONTRIBUTING.md [hunk 8]
```diff
@@ -102,6 +102,20 @@ Good commit message subject :
 Document how to install the project
 ```
 
+Also, when you specify what you did avoid commit message subjects with "Fix bug
+in such and such feature". Saying you are fixing something implies the previous
+implementation was wrong and yours is right, which might not be even true.
+Instead, humbly say what you did technically and **then** explain how this is
```
Comment (ju1ius) [inline doc/CONTRIBUTING.md:108]: I'm not sure this is worth saying.
When you propose a change, you want to fix a bug and you do this because you encountered the bug, so the previous implementation was indeed wrong in some way (or you think it was).
Or you refactor things because an implementation is wrong/incomplete/bloated/... (or you think it is).
Saying that you think you're right does not imply you're not humble, you're just stating your point.

IMO this is not the reason why writing commit messages this way is better.

File: doc/CONTRIBUTING.md [hunk 9]
```diff
@@ -102,6 +102,20 @@ Good commit message subject :
...
+call foo::bar() instead of bar::baz()
+
+This fixes a bug that arises when doing this or that, because baz() needs a
+flux capacitor object that might not be defined.
+Fixes #42
+```
+
 The description is optional but strongly recommended. It could be asked by the
 team if needed. PR will often lead to complicated, hard-to-read conversations
 with many links to other web pages.
```
Comment (ju1ius) [inline doc/CONTRIBUTING.md:121]: > The description is optional but strongly recommended.

Then why not requiring it ?

File: doc/CONTRIBUTING.md (hunk 8 above)
Comment (greg0ire) [inline doc/CONTRIBUTING.md:108]: > or you think it was

This is key. Sometimes you think it was, but the commit message for the commit that introduced what you are fixing is so imprecise that you don't see how the piece of code actually is useful, until you hit a bug.

Plus, the shortlog is what your doing, and thus stating _unquestionable facts_ (unlike I'm probalby fixing a bug). Fixing a bug is _why_ you are doing it, hence it should be in the § below. Plus, while it is strongly recommended to refer to an existing issue, you should always at least sum it up (still not relying on a web application to explain what you do). Summing the issue up will take more than 50 chars (hopefully). `Fixes #666` is a plus. It is a big one, but it should be considered just a plus.
Issues and PRs are confusing, long to read, can include gifs and emojis and a lot of noise. Your commit message should extract everything that is interesting in the PR/issue, so that the history gets it.

To sum up, the value should be in the git history, and not on github. On github, you speak with others to get help on creating the value.

File: doc/CONTRIBUTING.md (hunk 9 above)
Comment (greg0ire) [inline doc/CONTRIBUTING.md:121]: Because sometimes you just fix a typo. The why is so obvious it is optional.

Comment (greg0ire) [inline doc/CONTRIBUTING.md:121]: See my commit "fix grammar" in this PR

File: doc/CONTRIBUTING.md (hunk 8 above)
Comment (ju1ius) [inline doc/CONTRIBUTING.md:108]: I definitely totally agree on what you say.
Im' just saying that a phrase like the one you just did:

> the shortlog is what your doing, and thus stating unquestionable facts

...explains things _way_ better than the previous one (with the part about «being humble», which introduce a rather emotional language, that could be mistaken as defensive).

Comment (greg0ire) [inline doc/CONTRIBUTING.md:108]: Yeah that was uncalled for, I'll rework that. Thanks for helping me finding the words.

Comment (ju1ius) [inline doc/CONTRIBUTING.md:108]: Maybe add something like:

> commit messages and descriptions should stay technical.
> State facts, not opinions.
> Keep your opinions and emotions for the PR comments, but remember to be kind to one another.

Comment (greg0ire) [inline doc/CONTRIBUTING.md:108]: I'm leaving out the second part because I'm not writing a code of conduct right now, but this could be added somewhere else. See #2
